### PR TITLE
feat: support `display` prop for all components with styled props

### PR DESCRIPTION
**Pull request by chaitanyadeorukhkar**
fixes: #1164 

Components that didn't need any changes (display prop wasn't breaking anything):
- Button
- Checkbox, CheckboxGroup
- Radio, RadioGroup
- List
- Spinner
- Typography
- Link 
- Code(it breaks with certain display props but adding wrappers on typography would be very messy and its okay if people break it since it would be unusable in the broken state)

Not added display prop since it doesn't support any styled props
- Tooltip
- Bottomsheet
- Modal

Misc fixes in this PR:
- Added styledProps to Switch
- Added `testID` & meta attribute to Dropdown

Win 🧘‍♂️:
Screenshot 2023-07-13 at 4 41 42 PM

### CODE REVIEW

File: packages/blade/src/components/Alert/Alert.tsx [hunk 1]
```diff
@@ -293,29 +293,32 @@ const Alert = ({
   }
 
   return (
-    <StyledAlert
-      intent={intent}
-      contrastType={contrastType}
-      isFullWidth={isFullWidth}
-      isDesktop={isDesktop}
-      textAlign={'left' as never}
+    <BaseBox
```
Comment (saurabhdaware) [inline packages/blade/src/components/Alert/Alert.tsx:296]: I think all Box wrappers should have `display: inline-block` by default right? otherwise they'll take up full width even when the component inside it is inlined

File: packages/blade/src/components/Divider/Divider.stories.tsx [hunk 2]
```diff
@@ -133,13 +135,17 @@ const DividerVerticalTemplate: ComponentStory<typeof DividerComponent> = () => {
           flexDirection="row"
           height="50px"
         >
-          <Button variant="secondary" marginRight="spacing.4" isFullWidth>
-            Sign up
-          </Button>
+          <BaseBox display="flex" flex={1}>
```
Comment (saurabhdaware) [inline packages/blade/src/components/Divider/Divider.stories.tsx:138]: is this a related change?

File: packages/blade/src/components/Dropdown/Dropdown.tsx [hunk 3]
```diff
@@ -182,8 +184,13 @@ const _Dropdown = ({
   return (
     <BottomSheetAndDropdownGlueContext.Provider value={BottomSheetAndDropdownGlueContextValue}>
       <DropdownContext.Provider value={contextValue}>
-        <BaseBox position="relative" textAlign={'left' as never} {...getStyledProps(styledProps)}>
-          {children}
+        <BaseBox
```
Comment (saurabhdaware) [inline packages/blade/src/components/Dropdown/Dropdown.tsx:187]: is wrapper needed on Dropdown?

File: packages/blade/src/components/Alert/Alert.tsx (hunk 1 above)
Comment (chaitanyadeorukhkar) [reply packages/blade/src/components/Alert/Alert.tsx:296]: Better to keep it unset by default? Since we allow people to change the display prop on the wrapper, they can tweak it however they want. 

Maybe some specific components that needs to have a specific display prop, like `inline`, we can replicate that on the wrapper component by default but allow consumers to override it anyway. Let me see if I can find any such instances in this PR

File: packages/blade/src/components/Divider/Divider.stories.tsx (hunk 2 above)
Comment (chaitanyadeorukhkar) [reply packages/blade/src/components/Divider/Divider.stories.tsx:138]: Not really but story was breaking on RN so just fixed it

File: packages/blade/src/components/Dropdown/Dropdown.tsx (hunk 3 above)
Comment (chaitanyadeorukhkar) [reply packages/blade/src/components/Dropdown/Dropdown.tsx:187]: Yeah, without it Dropdown was breaking especially if you make it flex, the sub-components get messed up

File: packages/blade/src/components/Alert/Alert.tsx (hunk 1 above)
Comment (chaitanyadeorukhkar) [reply packages/blade/src/components/Alert/Alert.tsx:296]: Did this for amount, other places it seems to not need it. Can you check once?

Comment (saurabhdaware) [reply packages/blade/src/components/Alert/Alert.tsx:296]: Can we do it on Counter, Badge, Indicators too? components that are small in width and usually used along with text by default?

File: packages/blade/src/components/Dropdown/Dropdown.tsx (hunk 3 above)
Comment (saurabhdaware) [reply packages/blade/src/components/Dropdown/Dropdown.tsx:187]: Ohh do we want to consider the flex scenarios as well? I was thinking more along the lines of that components shouldn't break when we pass `block` or `inline-block` to it. Because while doing responsiveness, those are the props that usually people use to "show" an element.

E.g.

```jsx
<Badge display={{ base: 'none', m: 'block' }}>I am badge visible on desktop</Badge>
```

Things breaking when we pass "flex" should be fine right? those will break the UI itself so it will be noticeable and won't be shipped. E.g. I can't think of scenario where someone would want to make Dropdown a display flex.

Comment (chaitanyadeorukhkar) [reply packages/blade/src/components/Dropdown/Dropdown.tsx:187]: I think we'll have to consider all scenarios right? It shouldn't break the component's visuals. Imagine someone shipped this code without testing on smaller screen and it breaks the component's UI itself. 

```tsx
<Badge display={{ base: 'none', s: 'flex' }}>I am badge visible on desktop</Badge>
```

File: packages/blade/src/components/Alert/Alert.tsx (hunk 1 above)
Comment (chaitanyadeorukhkar) [reply packages/blade/src/components/Alert/Alert.tsx:296]: So Counter, Badge & Indicators have flex wrappers so it shouldn't be much of a problem even if they make it inline

Comment (chaitanyadeorukhkar) [reply packages/blade/src/components/Alert/Alert.tsx:296]: Counter, Badge, & Indicators have been flex since forever now. With this change, they can now make it inline. I think its better to keep the way it is for now without changing it. I will make the wrapper also flex so that it remains as it was before

Comment (chaitanyadeorukhkar) [reply packages/blade/src/components/Alert/Alert.tsx:296]: Especially because these components are used within Card, Modal & Bottomsheet. Don't want to impact it there by enforcing it as inline here

Comment (saurabhdaware) [reply packages/blade/src/components/Alert/Alert.tsx:296]: Oh actually we did try to make Badge inline-flex but I just noticed that we didn't change display property of the wrapper container https://github.com/razorpay/blade/pull/1254

We can do this separately then but we should make it inline-flex / inline-block as it was raised by our consumers as well https://github.com/razorpay/blade/issues/1241, https://razorpay.slack.com/archives/CMQ3RBHEU/p1685526734598399?thread_ts=1685526177.654729&cid=CMQ3RBHEU.

Lets take it up separately since this change is big in itself already.

Comment (kamleshchandnani) [reply packages/blade/src/components/Alert/Alert.tsx:296]: lets create an issue to track it?

Comment (chaitanyadeorukhkar) [reply packages/blade/src/components/Alert/Alert.tsx:296]: Saurabh already did #1447

File: packages/blade/src/components/Input/BaseInput/BaseInput.tsx [hunk 4]
```diff
@@ -649,6 +649,7 @@ export const BaseInput = React.forwardRef<HTMLInputElement, BaseInputProps>(
           justifyContent={isLabelLeftPositioned ? 'center' : undefined}
           alignItems={isLabelLeftPositioned ? 'center' : undefined}
           position="relative"
+          width="100%"
```
Comment (kamleshchandnani) [inline packages/blade/src/components/Input/BaseInput/BaseInput.tsx:652]: why this?

**Comment by kamleshchandnani** [comment]
@chaitanyadeorukhkar resolve conflicts?

File: packages/blade/src/components/Input/BaseInput/BaseInput.tsx (hunk 4 above)
Comment (chaitanyadeorukhkar) [reply packages/blade/src/components/Input/BaseInput/BaseInput.tsx:652]: Without this, the input will break if we change the display property

File: packages/blade/src/components/Badge/Badge.tsx [hunk 5]
```diff
@@ -124,40 +124,40 @@ const _Badge = ({
   } as const;
 
   return (
-    <StyledBadge
-      backgroundColor={backgroundColor}
-      size={size}
-      textAlign={'left' as never}
+    <BaseBox
+      display="flex"
```
[the remark this answers is not in the document]
Comment (chaitanyadeorukhkar) [reply packages/blade/src/components/Badge/Badge.tsx:128]: getStyledProps() will prioritize it. 

```tsx
<BaseBox
      display="flex"
      {...metaAttribute({ name: MetaConstants.Badge, testID })}
      {...getStyledProps(styledProps)} // user's display prop will be added from here
    >
      <StyledBadge backgroundColor={backgroundColor} size={size} textAlign={'left' as never}>
        <BaseBox
```

File: packages/blade/src/components/Amount/Amount.tsx [hunk 6]
```diff
@@ -251,32 +251,36 @@ const _Amount = ({
...
-        marginRight="spacing.1"
-        fontWeight={currencyWeight}
-        fontSize={currencyFontSize}
-        color={currencyColor}
-        as={isReactNative ? undefined : 'span'}
+      <BaseBox
+        paddingLeft="spacing.2"
+        paddingRight="spacing.2"
+        // @TODO: fix casting of platform types. currently they all become `never` type
+        display={isReactNative ? castNativeType('flex') : castWebType('inline-flex')}
```
[the remark this answers is not in the document]
Comment (chaitanyadeorukhkar) [reply packages/blade/src/components/Amount/Amount.tsx:262]: Making the display prop of new wrapper same as the earlier one's to ensure nothing breaks in existing implementations. See the code below this, it has this too:
```tsx
        display={isReactNative ? castNativeType('flex') : castWebType('inline-flex')}
```

[the remark this answers is not in the document]
Comment (kamleshchandnani) [reply packages/blade/src/components/Amount/Amount.tsx:262]: TODO things fixing for this file?

File: packages/blade/src/components/Accordion/Accordion.tsx [hunk 7]
```diff
@@ -95,14 +95,14 @@ const Accordion = ({
   return (
     <AccordionContext.Provider value={accordionContext}>
       <BaseBox
-        minWidth={MIN_WIDTH}
-        maxWidth={MAX_WIDTH}
         {...metaAttribute({ name: MetaConstants.Accordion, testID })}
```
[the remark this answers is not in the document]
Comment (kamleshchandnani) [reply packages/blade/src/components/Accordion/Accordion.tsx:98]: `metaAttribute` also needs to be passed to this `<BaseBox minWidth={MIN_WIDTH} maxWidth={MAX_WIDTH} width="100%">` as well?

File: packages/blade/src/components/Amount/Amount.tsx (hunk 6 above)
[the remark this answers is not in the document]
Comment (chaitanyadeorukhkar) [reply packages/blade/src/components/Amount/Amount.tsx:262]: Tried looking into it, might take a bit longer - its a weird issue. Lets not block the PR for it. We can look at it when working on some component where similar use-case would come up

File: packages/blade/src/components/Accordion/Accordion.tsx (hunk 7 above)
[the remark this answers is not in the document]
Comment (chaitanyadeorukhkar) [reply packages/blade/src/components/Accordion/Accordion.tsx:98]: No... the meta Attribute will be there only on the root level wrapper, right?

[the remark this answers is not in the document]
Comment (kamleshchandnani) [reply packages/blade/src/components/Accordion/Accordion.tsx:98]: no, after @anuraghazra fix we need to add it to all the nodes otherwise the coverage will mess up

File: .changeset/lovely-pants-breathe.md [hunk 8]
```diff
@@ -0,0 +1,7 @@
+---
+"@razorpay/blade": minor
+---
+
+feat: support `display` prop for all components
```
Comment (snitin315) [inline .changeset/lovely-pants-breathe.md:5]: This might be misleading.
```suggestion
feat: support `display` prop for all components with styled props
```

File: packages/blade/src/components/Accordion/Accordion.tsx (hunk 7 above)
[the remark this answers is not in the document]
Comment (chaitanyadeorukhkar) [reply packages/blade/src/components/Accordion/Accordion.tsx:98]: Its basebox, so we don't need to add anything. Confirmed with @anuraghazra

**Review by kamleshchandnani — APPROVED** [review]

**Review by kamleshchandnani — APPROVED** [review]
